test/e2e/util_gatewayapi.go

```diff
@@ -0,0 +1,605 @@
+	t.Helper()
+	crd := &apiextensionsv1.CustomResourceDefinition{}
+	name := types.NamespacedName{"", crdname}
+	crdVersion := ""
+
+	err := wait.PollImmediate(1*time.Second, 30*time.Second, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), name, crd); err != nil {
+			t.Logf("failed to get crd %s: %v", name, err)
+			return false, nil
+		}
+		crdConditions := crd.Status.Conditions
+		for _, version := range crd.Spec.Versions {
+			if version.Served {
+				crdVersion = version.Name
+			}
+		}
+		for _, c := range crdConditions {
+			if c.Type == apiextensionsv1.Established && c.Status == apiextensionsv1.ConditionTrue {
+				return true, nil
+			}
+		}
+		t.Logf("failed to find crd %s to be Established", name)
+		return false, nil
+	})
+	return crdVersion, err
+}
+
+// assertCanCreateHttpRoute  attempts to create a gatewayClass with the given namespace, name, parent namespace,
+// hostname, backend reference name, and gateway.
+// Returns false if it cannot create the httpRoute, otherwise returns true.
+func assertCanCreateHttpRoute(t *testing.T, ns, name, parentns, hostname, berefname string, gateway *gwapi.Gateway) (bool, *gwapi.HTTPRoute) {
+	t.Helper()
+
+	httproute, err := createHttpRoute(ns, name, parentns, hostname, berefname, gateway)
+	if err != nil {
+		t.Logf("error creating httpRoute: %v", err)
+		return false, nil
+	}
+	return true, httproute
+}
+
+// createHttpRoute checks if the HTTPRoute can be created.
+// If it can't an error is returned.
+func createHttpRoute(namespace, routename, parentnamespace, hostname, backendrefname string, gateway *gwapi.Gateway) (*gwapi.HTTPRoute, error) {
+	// Just in case gateway creation failed, supply a fake gateway name.
+	name := "NONE"
+	if gateway != nil {
+		name = gateway.Name
+	}
+
+	// Create the backend (service and pod) needed for the route to have resolvedRefs=true.
+	// The http route, service, and pod are cleaned up when the namespace is automatically deleted.
+	// buildEchoPod builds a pod that listens on port 8080.
+	echoPod := buildEchoPod(backendrefname, namespace)
+	if err := kclient.Create(context.TODO(), echoPod); err != nil {
+		return nil, fmt.Errorf("failed to create pod %s/%s: %v", namespace, echoPod.Name, err)
+	}
+	// buildEchoService builds a service that targets port 8080.
+	echoService := buildEchoService(echoPod.Name, namespace, echoPod.ObjectMeta.Labels)
+	if err := kclient.Create(context.TODO(), echoService); err != nil {
+		return nil, fmt.Errorf("failed to create service %s/%s: %v", echoService.Namespace, echoService.Name, err)
+	}
+
+	httpRoute := buildHTTPRoute(routename, namespace, name, parentnamespace, hostname, backendrefname)
+	if err := kclient.Create(context.TODO(), httpRoute); err != nil {
+		if kerrors.IsAlreadyExists(err) {
+			name := types.NamespacedName{namespace, routename}
+			if err = kclient.Get(context.TODO(), name, httpRoute); err == nil {
+				return httpRoute, nil
+			}
```

Comment:
It's highly unlikely that you'd get an "already exists" error and also get an error on the `Get` call to the same resource, but still, that error isn't handled here

Reply:
Fixed.